### PR TITLE
Allow list builder to change type

### PR DESCRIPTION
**Pull request by patszt**
In this PR we allow the some form of a common type casting inside a records and lists.

Previously changing a field inside a nested record was considered as a casting of a `record_type` so it was not implemented.
Same with the lists. A `list<int>` changing to `list<string>` is a list casting, which was not planned to be implemented yet.

It seems that the live data has a lot of `ip` -> `string` conversions that must be supported. These conversions occurr inside the lists and inside the nested record so this type of casting should be supported.

In this PR we allow:
- changing of a list_type when the added value requires the common type casting. E.g when we have a `list<ip>` we may add a "www.tenzir.com". This is inferred as a string type and the whole list must be cast to a `list<string>` in order to be compatible with the new string input.
- In case of `list<record<...>>` the field inside a record can also require the common type casting. E.g `list<record<ip_field: ip>>` can be converted to `list<record<ip_field: string>>` when we try to add some not castable data to the `ip_field`

### CODE REVIEW

File: libtenzir/include/tenzir/cast.hpp [hunk 1]
```diff
@@ -261,6 +293,45 @@ struct cast_helper<list_type, list_type> {
              from_array->null_bitmap(), from_array->null_count())
       .ValueOrDie();
   }
+
+  static auto cast_value(const list_type& from_type, const list& value,
+                         const list_type& to_type) noexcept
+    -> caf::expected<type_to_data_t<list_type>> {
+    if (from_type == to_type)
+      return value;
+    auto output = to_type.construct();
```
Comment (dominiklohmann) [inline libtenzir/include/tenzir/cast.hpp:302]:
```suggestion
    output.reserve(value.size());
```

File: libtenzir/include/tenzir/cast.hpp [hunk 2]
```diff
@@ -335,6 +406,53 @@ struct cast_helper<record_type, record_type> {
...
+        if (not index) {
+          // The field does not exist, so we insert a null.
+          ret[to_field.name] = caf::none;
+          continue;
+        }
+        // The field exists, so we can insert the cast column.
+        auto input_at_path = descend(std::addressof(in), key);
+        if (not input_at_path)
+          return std::move(input_at_path.error());
+        if (**input_at_path == caf::none) {
```
Comment (dominiklohmann) [inline libtenzir/include/tenzir/cast.hpp:441]: Please add an assertion that it is safe to dereference twice here.

File: libtenzir/include/tenzir/cast.hpp [hunk 3]
```diff
@@ -702,6 +820,36 @@ struct cast_helper<double_type, enumeration_type> {
...
+    if (can)
+      return value;
+    return std::move(can.error());
+  }
+
+  static auto cast(const enumeration_type&,
+                   std::shared_ptr<type_to_arrow_array_t<enumeration_type>>,
+                   const enumeration_type&) noexcept
+    -> std::shared_ptr<type_to_arrow_array_t<enumeration_type>> {
+    die("enumeration to enumeration cast unimplemented");
```
Comment (dominiklohmann) [inline libtenzir/include/tenzir/cast.hpp:849]:
```suggestion
    TENZIR_ASSERT(can_cast(from, to));
    return array;
```

File: libtenzir/include/tenzir/cast.hpp [hunk 4]
```diff
@@ -922,26 +1070,30 @@ auto cast_value(const FromType& from_type, ValueType&& value,
...
       auto status = ret->AppendNull();
       TENZIR_ASSERT(status.ok());
       continue;
     }
-    auto converted = cast_value(from_type, *v, to_type);
-    if (not converted)
-      return converted.error();
-    if constexpr (not std::is_same_v<decltype(converted), caf::expected<void>>) {
-      auto status = append_builder(to_type, *ret, *converted);
+    auto converted = cast_value(from_type, materialize(*v), to_type);
```
Comment (dominiklohmann) [inline libtenzir/include/tenzir/cast.hpp:1085]: This should not need to materialize. The casting API should accept views as arguments. I think this is a small change to make, as it mostly affects records and lists.

File: libtenzir/include/tenzir/detail/adaptive_table_slice_builder_guards.hpp [hunk 5]
```diff
@@ -25,6 +25,9 @@ struct type_from_data
 template <class T>
 using type_from_data_t = typename type_from_data<T>::type;
 
+using parent_record_builder_provider
+  = std::function<concrete_series_builder<record_type>*()>;
```
Comment (dominiklohmann) [inline libtenzir/include/tenzir/detail/adaptive_table_slice_builder_guards.hpp:29]: Please document the contract of this type.

File: libtenzir/include/tenzir/detail/adaptive_table_slice_builder_guards.hpp [hunk 6]
```diff
@@ -129,26 +133,34 @@ class list_guard {
...
-                                           decltype(cast_val)>) {
-            append_value_to_builder(t, *cast_val);
+          if (cast_val) {
+            // "this" is sometimes not used due to the if constexpr below.
+            // It prevents compilation warning treated as an error.
+            (void)this;
+            // We don't handle caf::expected<void> here because the cast api
+            // only returns caf::expected<void> if the operation is not
+            // supported. The caf::expected<void> will fall into common_type
+            // casting case.
```
Comment (dominiklohmann) [inline libtenzir/include/tenzir/detail/adaptive_table_slice_builder_guards.hpp:152]: This seems brittle to me from an API user perspective. I would expect a cast to type T to return me an `expected<type_to_data_t<T>>`, regardless of whether that cast is possible or not.

**Review by dominiklohmann — COMMENTED** [review]
A first look, will do a second pass later.

File: libtenzir/src/detail/series_builders.cpp [hunk 7]
```diff
@@ -8,8 +8,260 @@
...
+          TENZIR_ASSERT(status.ok());
+        }
+        return should_append_builder;
+      },
+      [](auto&&...) -> bool {
+        TENZIR_WARN("unexpected data/builder/type");
+        return true;
+      },
+    },
+    type, builder, data);
```
[the remark this answers is not in the document]
Comment (patszt) [reply libtenzir/src/detail/series_builders.cpp]: done in 3a3ac78a90fc52d89c6988be523478a0a4203c9e

File: libtenzir/include/tenzir/cast.hpp (hunk 1 above)
Comment (patszt) [reply libtenzir/include/tenzir/cast.hpp:285]: done in 75dbddf6033f4098ed572de5ec48384401897aef

File: libtenzir/include/tenzir/cast.hpp (hunk 2 above)
Comment (patszt) [reply libtenzir/include/tenzir/cast.hpp:426]: done in 12dd25984f505a5630e02fbf12bb0785043bba53

File: libtenzir/include/tenzir/detail/adaptive_table_slice_builder_guards.hpp (hunk 5 above)
Comment (patszt) [reply libtenzir/include/tenzir/detail/adaptive_table_slice_builder_guards.hpp:35]: done in ce71c24dd38c9e8d377d0c6c42b1c6c5bbc1f165

File: libtenzir/include/tenzir/detail/adaptive_table_slice_builder_guards.hpp (hunk 6 above)
Comment (patszt) [reply libtenzir/include/tenzir/detail/adaptive_table_slice_builder_guards.hpp]: done in d19a6661b1904fd4398fd4caac06f2caa10c5bd3

File: libtenzir/include/tenzir/cast.hpp (hunk 4 above)
Comment (patszt) [reply libtenzir/include/tenzir/cast.hpp]: done in 50991a780bcf1e4322d07b81e8cecd5f56a63951

**Review by dominiklohmann — APPROVED** [review]
Thanks for incorporating the feedback

**Pull request merged**